mlir/lib/Dialect/Presburger/Set.cpp

```diff
@@ -0,0 +1,396 @@
+
+  return coeffs;
+}
+
+SmallVector<int64_t, 64> complementIneq(const ArrayRef<int64_t> &ineq) {
+  return inequalityFromEquality(ineq, true, true);
+}
+// Return the set difference B - S and accumulate the result into `result`.
+// `simplex` must correspond to B.
+//
+// In the following, U denotes union, /\ denotes intersection, - denotes set
+// subtraction and ~ denotes complement.
+// Let B be the basic set and S = (U_i S_i) be the set. We want B - (U_i S_i).
+//
+// Let S_i = /\_j S_ij. To compute B - S_i = B /\ ~S_i, we partition S_i based
+// on the first violated constraint:
+// ~S_i = (~S_i1) U (S_i1 /\ ~S_i2) U (S_i1 /\ S_i2 /\ ~S_i3) U ...
+// And the required result is (B /\ ~S_i1) U (B /\ S_i1 /\ ~S_i2) U ...
+// We recurse by subtracting U_{j > i} S_j from each of these parts and
+// returning the union of the results.
+//
+// TODO reimplement this heuristic:
+// As a heuristic, we try adding all the constraints and check if simplex
+// says that the intersection is empty. Also, in the process we find out that
+// some constraints are redundant, which we then ignore.
+void subtractRecursively(FlatAffineConstraints &B, Simplex &simplex,
+                         const PresburgerSet &S, unsigned i,
+                         PresburgerSet &result) {
+  if (i == S.getNumBasicSets()) {
+    FlatAffineConstraints BCopy = B;
+    // BCopy.simplify();
+    result.addFlatAffineConstraints(std::move(BCopy));
+    return;
+  }
+  const FlatAffineConstraints &S_i = S.getFlatAffineConstraints()[i];
+  auto initialSnap = simplex.getSnapshot();
+  // unsigned offset = simplex.numberConstraints();
+  // simplex.addBasicSetAsInequalities(set_i);
+  simplex.addFlatAffineConstraints(S_i);
+
+  if (simplex.isEmpty()) {
+    simplex.rollback(initialSnap);
+    subtractRecursively(B, simplex, S, i + 1, result);
+    return;
+  }
+
+  /*std::vector<bool> isMarkedRedundant;
+  for (size_t j = 0; j < 2 * S_i.getNumEqualities() + S_i.getNumInequalities();
+       j++)
+    isMarkedRedundant.push_back(simplex.isMarkedRedundant(offset + j));
+  */
+  simplex.rollback(initialSnap);
+  // TODO benchmark does it make a lot of difference if we always_inline this?
+  auto addInequalityFromEquality = [&](const ArrayRef<int64_t> &eq,
+                                       bool negated, bool strict) {
+    SmallVector<int64_t, 64> coeffs =
+        inequalityFromEquality(eq, negated, strict);
+    B.addInequality(coeffs);
+    simplex.addInequality(coeffs);
+  };
+  auto recurseWithInequalityFromEquality = [&, i](const ArrayRef<int64_t> &eq,
+                                                  bool negated, bool strict) {
+    size_t snap = simplex.getSnapshot();
+    addInequalityFromEquality(eq, negated, strict);
+
+    subtractRecursively(B, simplex, S, i + 1, result);
+
+    // TODO check if this removes the right inequality
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+  };
+
+  size_t addedIneqs = 0;
```

Comment:
Might be cleaner to store the number of inequalities present at the start and remove all the ones added beyond that later, instead of keeping track of how many were added.

Reply:
Correct. Furthermore it should be of type `unsigned`, not `size_t`.